File: internal/core/permissions.go
package core

// Permission is an explicit capability granted by a panel role. Keeping the
// policy in the core package makes the API and other adapters share the same
// authorization vocabulary instead of relying on route-specific rank checks.
type Permission string

const (
	PermissionOverviewRead     Permission = "overview.read"
	PermissionAgentsRead       Permission = "agents.read"
	PermissionAgentsManage     Permission = "agents.manage"
	PermissionClientAccessRead Permission = "client-access.read"
	PermissionDeploymentsRead  Permission = "deployments.read"
	PermissionCatalogsRead     Permission = "catalogs.read"
	PermissionAgentConfigRead  Permission = "agent-config.read"
	PermissionAgentConfigWrite Permission = "agent-config.write"
	PermissionConfigsRead      Permission = "configs.read"
	PermissionConfigsWrite     Permission = "configs.write"
	PermissionConfigsDelete    Permission = "configs.delete"
	PermissionConfigsRestore   Permission = "configs.restore"
	PermissionTasksRead        Permission = "tasks.read"
	PermissionTasksExecute     Permission = "tasks.execute"
	PermissionEnrollmentManage Permission = "enrollment.manage"
	PermissionSettingsRead     Permission = "settings.read"
	PermissionSettingsManage   Permission = "settings.manage"
	PermissionAuditRead        Permission = "audit.read"
	PermissionMetricsRead      Permission = "metrics.read"
	PermissionPanelMetricsRead Permission = "panel-metrics.read"
	PermissionCoreLogsRead     Permission = "core-logs.read"
	PermissionTrafficRead      Permission = "traffic.read"
	PermissionTrafficManage    Permission = "traffic.manage"
	PermissionUsersManage      Permission = "users.manage"
	PermissionTemplatesRead    Permission = "templates.read"
	PermissionTemplatesWrite   Permission = "templates.write"
	PermissionTemplatesDelete  Permission = "templates.delete"
)

var allPermissions = []Permission{
	PermissionOverviewRead, PermissionAgentsRead, PermissionAgentsManage,
	PermissionClientAccessRead, PermissionDeploymentsRead, PermissionCatalogsRead,
	PermissionAgentConfigRead, PermissionAgentConfigWrite, PermissionConfigsRead,
	PermissionConfigsWrite, PermissionConfigsDelete, PermissionConfigsRestore,
	PermissionTasksRead, PermissionTasksExecute, PermissionEnrollmentManage,
	PermissionSettingsRead, PermissionSettingsManage, PermissionAuditRead,
	PermissionMetricsRead, PermissionPanelMetricsRead, PermissionCoreLogsRead, PermissionTrafficRead, PermissionTrafficManage, PermissionUsersManage, PermissionTemplatesRead,
	PermissionTemplatesWrite, PermissionTemplatesDelete,
}

// These sets are built once during package initialization and never exposed
// for mutation. Login/session reads can share them without rebuilding policy.
var (
	rolePermissions        = map[Role]map[Permission]struct{}{}
	grantablePermissions   []Permission
	grantablePermissionSet map[Permission]struct{}
)

func AllPermissions() []Permission { return append([]Permission(nil), allPermissions...) }

// PermissionGrantsAdministration identifies administrator-only capabilities.
// Explicit user grants must agree with the store's existing administrator
// scope instead of implying authority that the stored role does not have.
func PermissionGrantsAdministration(permission Permission) bool {
	return permission == PermissionUsersManage
}

// GrantablePermissions returns the capabilities an administrator may assign
// to a non-administrator. Store writes reject administrator-only requests;
// reads use this allowlist to filter impossible grants in older backups.
func GrantablePermissions() []Permission {
	return append([]Permission(nil), grantablePermissions...)
}

// NormalizePermissions keeps known capabilities in their original order,
// without duplicates. Account writes must also validate the resulting role.
func NormalizePermissions(values []Permission) []Permission {
	return normalizePermissions(values, rolePermissions[RoleAdmin])
}

// NormalizeGrantablePermissions also removes administrator-only capabilities
// when resolving an existing non-administrator account.
func NormalizeGrantablePermissions(values []Permission) []Permission {
	return normalizePermissions(values, grantablePermissionSet)
}

func normalizePermissions(values []Permission, allowed map[Permission]struct{}) []Permission {
	capacity := min(len(values), len(allowed))
	seen := make(map[Permission]struct{}, capacity)
	result := make([]Permission, 0, capacity)
	for _, value := range values {
		if _, ok := allowed[value]; !ok {
			continue
		}
		if _, ok := seen[value]; ok {
			continue
		}
		seen[value] = struct{}{}
		result = append(result, value)
	}
	return result
}

// Admin is intentionally built from the union of every declared capability.
// New capabilities must be listed here so the administrator remains the
// complete break-glass role while lower roles stay deny-by-default.
func init() {
	rolePermissions[RoleAdmin] = permissionSet(allPermissions...)
	for _, permission := range allPermissions {
		if !PermissionGrantsAdministration(permission) {
			grantablePermissions = append(grantablePermissions, permission)
		}
	}
	grantablePermissionSet = permissionSet(grantablePermissions...)
}

func permissionSet(values ...Permission) map[Permission]struct{} {
	result := make(map[Permission]struct{}, len(values))
	for _, value := range values {
		result[value] = struct{}{}
	}
	return result
}

// Allows reports whether this role has the named capability. Unknown roles
// and capabilities are denied by default.
func (role Role) Allows(permission Permission) bool {
	_, ok := rolePermissions[role][permission]
	return ok
}

func HasPermission(values []Permission, wanted Permission) bool {
	for _, value := range values {
		if value == wanted {
			return true
		}
	}
	return false
}

File: frontend/modules/permissions.js
const rolePermissions = {
  admin: new Set(["overview.read", "agents.read", "agents.manage", "client-access.read", "deployments.read", "catalogs.read", "agent-config.read", "agent-config.write", "configs.read", "configs.write", "configs.delete", "configs.restore", "tasks.read", "tasks.execute", "enrollment.manage", "settings.read", "settings.manage", "audit.read", "metrics.read", "traffic.read", "traffic.manage", "users.manage", "core-logs.read", "templates.read", "templates.write", "templates.delete"]),
  operator: new Set(["overview.read", "agents.read", "deployments.read", "client-access.read", "catalogs.read", "agent-config.read", "agent-config.write", "configs.read", "configs.write", "tasks.read", "tasks.execute", "settings.read", "audit.read", "metrics.read", "traffic.read", "traffic.manage", "templates.read", "templates.write", "core-logs.read"]),
  auditor: new Set(["overview.read", "agents.read", "deployments.read", "tasks.read", "settings.read", "audit.read", "metrics.read", "traffic.read", "core-logs.read"]),
  readonly: new Set(["overview.read", "agents.read", "deployments.read", "client-access.read", "catalogs.read", "agent-config.read", "configs.read", "tasks.read", "settings.read", "audit.read", "metrics.read", "traffic.read", "templates.read", "core-logs.read"]),
};
const roleRanks = { readonly: 1, auditor: 1, operator: 2, admin: 3 };

export function createPermissionChecker(state) {
  const can = (capability, agent) => {
    const role = state.session?.role;
    if (role === "admin") return true;
    if (capability === "agent-access.read") return Boolean(state.session);
    if (capability === "users.manage") return false;
    if (agent && ["host.manage", "operator", "agents.manage", "traffic.manage", "enrollment.manage"].includes(capability) && (agent.can_manage === false || (role === "user" && agent.can_manage !== true))) return false;
    if (capability === "host.manage") return can("tasks.execute");
    if (capability === "system-bbr.read") return can("agents.read");
    if (capability === "operator") capability = "tasks.execute";
    if (role === "user") return (state.session?.permissions || []).includes(capability);
    if (capability in roleRanks) return (roleRanks[role] || 0) >= (roleRanks[capability] || 0);
    return Boolean(rolePermissions[role]?.has(capability));
  };
  return can;
}
